sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemSchema.java

```diff
@@ -0,0 +1,655 @@
+      //auth check for available segments
+      final Iterator<Entry<DataSegment, SegmentMetadataHolder>> authorizedAvailableSegments = getAuthorizedAvailableSegments(
+          availableSegmentEntries,
+          root
+      );
+
+      final FluentIterable<Object[]> availableSegments = FluentIterable
+          .from(() -> authorizedAvailableSegments)
+          .transform(val -> {
+            try {
+              if (!availableSegmentIds.contains(val.getKey().getIdentifier())) {
+                availableSegmentIds.add(val.getKey().getIdentifier());
+              }
+              return new Object[]{
+                  val.getKey().getIdentifier(),
+                  val.getKey().getDataSource(),
+                  val.getKey().getInterval().getStart(),
+                  val.getKey().getInterval().getEnd(),
+                  val.getKey().getSize(),
+                  val.getKey().getVersion(),
+                  val.getKey().getShardSpec().getPartitionNum(),
+                  val.getValue().getNumReplicas(),
+                  val.getValue().getNumRows(),
+                  val.getValue().isPublished(),
+                  val.getValue().isAvailable(),
+                  val.getValue().isRealtime(),
+                  jsonMapper.writeValueAsString(val.getKey())
+              };
+            }
+            catch (JsonProcessingException e) {
+              throw new RuntimeException(StringUtils.format(
+                  "Error getting segment payload for segment %s",
+                  val.getKey().getIdentifier()
+              ), e);
+            }
+          });
+
+      //auth check for published segments
+      final CloseableIterator<DataSegment> authorizedPublishedSegments = getAuthorizedPublishedSegments(
+          metadataSegments,
+          root
+      );
+      final FluentIterable<Object[]> publishedSegments = FluentIterable
+          .from(() -> authorizedPublishedSegments)
+          .transform(val -> {
+            try {
+              if (availableSegmentIds.contains(val.getIdentifier())) {
+                return null;
+              }
+              return new Object[]{
+                  val.getIdentifier(),
+                  val.getDataSource(),
+                  val.getInterval().getStart(),
+                  val.getInterval().getEnd(),
+                  val.getSize(),
+                  val.getVersion(),
+                  val.getShardSpec().getPartitionNum(),
+                  0L,
+                  -1L,
+                  1L,
+                  0L,
+                  0L,
+                  jsonMapper.writeValueAsString(val)
+              };
+            }
+            catch (JsonProcessingException e) {
+              throw new RuntimeException(StringUtils.format(
+                  "Error getting segment payload for segment %s",
+                  val.getIdentifier()
+              ), e);
+            }
+          });
+      try {
+        authorizedPublishedSegments.close();
```

Comment:
`authorizedPublishedSegments` is an iterator and the transform of `publishedSegments` is executed lazily. So, closing `authorizedPublishedSegments` here means that `authorizedPublishedSegments` is closed even before starting the iteration. 

`authorizedPublishedSegments.close()` should be called when the iteration on `Linq4j.asEnumerable(allSegments).where(t -> t != null)` is finished. But, I'm not sure what the best way is. @gianm any idea?

Reply:
Yes, i was not totally sure where to close `authorizedPublishedSegments`. I'll remove it from here and check what is the right place.